server/src/main/scala/com/cloudera/livy/server/interactive/InteractiveSession.scala

```diff
@@ -351,6 +348,21 @@ class InteractiveSession(
   _appId = appIdHint
   sessionStore.save(RECOVERY_SESSION_TYPE, recoveryMetadata)
 
+  // TODO Replace this with a Rpc call from repl to server.
+  private val stateThread = new Thread(new Runnable {
+    override def run(): Unit = {
+      try {
+        while (_state.isActive) {
+          // State is also updated when we get statement results from repl, not just here.
+          setSessionStateFromReplState(client.map(_.getReplState.get()))
+          Thread.sleep(1000)
+        }
+      } catch {
+        case _: InterruptedException =>
+      }
+    }
+  })
+
```

Comment:
I increased the timeout to 30s. Ideally it should be a LivyConf. But this doesn't really matter since I will submit a following PR to remove this thread.